tempest-testing-internal/src/main/kotlin/app/cash/tempest/testing/internal/TestDynamoDbService.kt

```diff
@@ -46,21 +46,29 @@ class TestDynamoDbService private constructor(
 
   companion object {
     private val defaultPorts = ConcurrentHashMap<String, Int>()
-    private fun defaultPort(key: String): Int {
+    private fun defaultPort(key: String): PortHolder {
+      var releasePort: () -> Unit = {}
       // Only pick random port once to share one test server with multiple tests.
-      return defaultPorts.getOrPut(key, ::pickRandomPort)
+      val port = defaultPorts.getOrPut(key) {
+        val socket = allocateRandomPort()
+        releasePort = { socket.close() }
+        socket.localPort
+      }
+      return PortHolder(port, releasePort)
     }
 
     private val runningServers = ConcurrentHashMap.newKeySet<String>()
     private val log = getLogger<TestDynamoDbService>()
 
     @JvmStatic
     fun create(serverFactory: TestDynamoDbServer.Factory<*>, tables: List<TestTable>, port: Int? = null): TestDynamoDbService {
-      val port = port ?: defaultPort(serverFactory.toString())
+      val portHolder = port?.let { PortHolder(it) } ?: defaultPort(serverFactory.toString())
       return TestDynamoDbService(
-        DefaultTestDynamoDbClient(tables, port),
-        serverFactory.create(port)
+        DefaultTestDynamoDbClient(tables, portHolder.value),
+        serverFactory.create(portHolder.value, portHolder.releasePort)
       )
     }
   }
+
+  private data class PortHolder(val value: Int, val releasePort: () -> Unit = {})
```

Comment:
Again, the name `portHolder` is deceptive, except for its one implementor. I think this would be better as a class method, and use an `init` block to create the closure state for the socket.

Reply:
So sorry, I meant to highlight `releasePort` again, not `PortHolder`. `releasePort` describes an action, which this lambda does not necessarily perform. Without taking this action, the name of the class doesn't make sense. What I was proposing is something like this

```kotiln
data class PortHolder(val port: Int) {
    val socket: ServerSocket
    init {
        socket = ServerSocket(port)
    }
    
    fun release() {
        socket.close()
    }
}
```

Of course, this only works if you can create the socket when you create the class. That might not work everywhere. The idea was just to make `releasePort` do what it says on the tin.

Reply:
I'd like to tinker with this, but I can't run these tests locally. The failure appears to be an apple silicon compatibility issue; this is using the old tempest v1 kit, and since the tests pass in CI I am not sure I want to risk bumping them and affecting people pinned on v1 for whatever reason.

Since I can't tinker with this easily, and what you have works, and as you point out the class is private and not used outside this file, I think this is good enough.